### PR TITLE
Use jemalloc to compile dist on CI (x86_64-linux)

### PR DESCRIPTION
**Pull request by blyxyas**
Based on [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/Is.20the.20BOLT.20on.20CI.20using.20jemalloc/near/424424207)

I added Jemalloc for building LLVM, this should cut some CI time.

r? @Kobzol

### CODE REVIEW

File: src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile [hunk 1]
```diff
@@ -33,12 +33,15 @@ RUN yum upgrade -y && \
...
       zlib-devel.x86_64 \
+      jemalloc-devel
       && yum clean all
 
 RUN mkdir -p /rustroot/bin
 
 ENV PATH=/rustroot/bin:$PATH
 ENV LD_LIBRARY_PATH=/rustroot/lib64:/rustroot/lib32:/rustroot/lib
+# This is a RUN command because we need to do some command interpolation
+RUN export LD_PRELOAD=`jemalloc-config --libdir`/libjemalloc.so.`jemalloc-config --revision`
```
Comment (Kobzol) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile:44]: `RUN export` sadly [doesn't export](https://stackoverflow.com/a/33379487/1107768) the environment to following layers :( So this is basically a no-op. You either have to use `ENV` or run this command at the beginning of the script where you want it to be applied.

Comment (blyxyas) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile]: Hmm I think I've fixed it in the new commit.

Comment (Kobzol) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile]: The build clang script is actually only building the host LLVM, this build is fully cached, so there's no point in optimizing it :sweat_smile:. You'd need to modify the opt-dist command invocation.

Comment (blyxyas) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile]: :sweat_smile: Okis, I'll get to it. Thanks for the help!

Comment (blyxyas) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile]: I've moved the env var to the `ENV SCRIPT (opt-dist)`, the string that gets run to optimize. I think this is correct :)

File: src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile [hunk 2]
```diff
@@ -33,6 +33,7 @@ RUN yum upgrade -y && \
       xz \
       zlib-devel.i686 \
       zlib-devel.x86_64 \
+      jemalloc-devel
```
Comment (lqd) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile:36]:
```suggestion
      jemalloc-devel \
```

Comment (blyxyas) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile]: I'm such a goober.

**Review by Kobzol — COMMENTED** [review]
I'm pretty sure that this only applies to the first command, and not to the whole chain (commands after `&&`) :laughing:. Bash is annoying.

**Comment by Kobzol** [comment]
Also for the next attempt, we don't need to run the benchmarks, just try :)

**Comment by blyxyas** [comment]
Hmmm... How should I check if this implies an improvement :sweat_smile: The action took 2h 24m, we can compare that with an arbitrary merge, and compared to e.g. #122054, this one seems to take an 40 minutes (1h 44m), is there any other number that I can check? I don't think that installing a simple jemalloc dependency could increase that much CI time (it should cut some minutes.)

**Comment by Kobzol** [comment]
We have to do two (or more) try builds in a row, so that the Docker cache kicks in.

@bors try

**Comment by Kobzol** [comment]
Ah, the environment command is still wrong. We need to use it for the invocation of `opt-dist`, not for its build.
So it should look something like this:
```bash
ENV SCRIPT python3 ../x.py build --set rust.debug=true opt-dist && \
    LD_PRELOAD=`jemalloc-config --libdir`/libjemalloc.so.`jemalloc-config --revision` ./build/$HOSTS/stage0-tools-bin/opt-dist linux-ci -- python3 ../x.py dist \
    --host $HOSTS --target $HOSTS \
    --include-default-paths \
```

**Comment by blyxyas** [comment]
Sorry for wasting so much time. It's a little bit embarrassing to mess up a simple 2-line change so many times.

**Comment by Kobzol** [comment]
Nah, you're not wasting time, this could actually help in the end :) You should check out some of my PRs that modify CI workflows, I have run try builds several hundred times at least :laughing: It's normal.

@bors try

**Comment by Kobzol** [comment]
Hmm, this prints an empty line:
```bash
FOO=BAR echo $(printenv) | grep FOO
```
so that would suggest that `<key>=<val> command` doesn't propagate the env to subprocesses. But that's maybe just some shell specialty, for normal processes it should hopefully work, since if the parent process has the env, it should propagate it to children by default (I hope).

**Comment by tgross35** [comment]
> Hmm, this prints an empty line:
> 
> ```shell
> FOO=BAR echo $(printenv) | grep FOO
> ```
> 
> so that would suggest that `<key>=<val> command` doesn't propagate the env to subprocesses. But that's maybe just some shell specialty, for normal processes it should hopefully work, since if the parent process has the env, it should propagate it to children by default (I hope).

I think subshells only get `export`ed environment, `set -a` makes that the default for assignments. Regular processes get the inline environment though, e.g. `FOO=BAR sh -c 'printenv' | grep FOO` works

**Comment by Kobzol** [comment]
Ah, makes sense. Yeah, for a Rust process, it should be enough if the parent gets the env, and then it should (hopefully) propagate through everything (I think that bootstrap does indeed propagate ENVs by default).

> Feels like I learn something new about bash every day :)

Yeah, same :laughing:

File: src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile [hunk 3]
```diff
@@ -33,6 +33,7 @@ RUN yum upgrade -y && \
       xz \
       zlib-devel.i686 \
       zlib-devel.x86_64 \
+      jemalloc-devel \
```
Comment (cuviper) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile:36]: You don't need `-devel` -- the base package includes a `jemalloc.sh` wrapper script that will set `LD_PRELOAD` for you.

**Comment by Kobzol** [comment]
Doesn't seem to have helped. It's hard to say whether it was used at all though. Maybe we could check e.g. in bootstrap.

**Comment by blyxyas** [comment]
I'll do some debugging in my machine to make sure that it is 100% being compiled with jemalloc.

**Comment by Kobzol** [comment]
Inside the `x64` dist Linux Docker, an old version of GLIBC (2.17, I think) is used, to make sure that the built artifacts are compatible with such old GLIBC distributions.

File: src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile (hunk 3 above)
Comment (blyxyas) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile:36]: Sorry for the delay, do you know where this jemalloc.sh script is located?

Comment (cuviper) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile:36]: It's `/usr/bin/jemalloc.sh`, so that's even in the `PATH`.

**Comment by blyxyas** [comment]
Hmmmm... I've been trying for a while (for like a month and a half) to just get this to run in my machine, including changing distro and getting a new hard drive (and changing from HDD to SSD). I'll close this to get the issue free for people that have the workflow already set up. It seems like my PC is just not having it :/

**Pull request closed without merging**